### PR TITLE
[RFR] Use hooks in controllers

### PR DESCRIPTION
**Pull request by djhi**
I believe the first step before migrating our controller components to custom hooks is to make them use hooks in the first place.

- [x] `CreateController`
- [x] `EditController`
- [x] `ShowController`
- [x] `ListController`
- [x] Unit test the utility functions which have been extracted

### CODE REVIEW

File: packages/ra-core/src/controller/CreateController.tsx [hunk 1]
```diff
@@ -36,11 +36,22 @@ interface Props {
     resource: string;
 }
 
-interface EnhancedProps {
-    crudCreate: Dispatch<typeof crudCreateAction>;
-    isLoading: boolean;
-    translate: Translate;
-}
+export const getRecord = ({ state, search }, record: any = {}) =>
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/CreateController.tsx]: this should live in the reducers directory as it's a selector (same with the test of this function)

File: packages/ra-core/src/controller/CreateController.tsx [hunk 2]
```diff
@@ -83,95 +94,57 @@ interface EnhancedProps {
...
+        (state: ReduxState) => state.admin.loading > 0
+    );
 
-    constructor(props) {
-        super(props);
-        const {
-            location: { state, search },
-            record,
-        } = this.props;
-        this.record =
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/CreateController.tsx]: I think you've broken the clone feature by not porting this

File: packages/ra-core/src/controller/EditController.tsx [hunk 3]
```diff
@@ -91,123 +80,77 @@ interface EnhancedProps {
...
-            children,
-            id,
-            isLoading,
-            record,
-            resource,
-            translate,
-            version,
-        } = this.props;
+    useEffect(() => {
+        dispatch(resetForm(REDUX_FORM_NAME));
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/EditController.tsx]: resetForm wasn't called on mount in the previous version. Isn't there a consequence of calling it here?

File: packages/ra-core/src/controller/CreateController.tsx (hunk 1 above)
Comment (djhi) [reply packages/ra-core/src/controller/CreateController.tsx]: Agreed, I will extract selectors

File: packages/ra-core/src/controller/CreateController.tsx (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/controller/CreateController.tsx]: The e2e tests disagree with you ;)

File: packages/ra-core/src/controller/EditController.tsx (hunk 3 above)
Comment (djhi) [reply packages/ra-core/src/controller/EditController.tsx]: I think we have e2e tests for this case but I will check

File: packages/ra-core/src/controller/CreateController.tsx (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/controller/CreateController.tsx]: And btw I ported this. This is the `getRecord` function you commented earlier

**Comment by kopax** [comment]
This looks cool thanks! Will we have a part about what are controllers in the documentation (RFC style)?

**Comment by djhi** [comment]
> Will we have a part about what are controllers in the documentation (RFC style)?

See the roadmap ;)

**Comment by fzaninotto** [comment]
Awesome!

**Comment by djhi** [comment]
> Awesome!
:hugs: 

Now we can extract custom hooks :rocket: !

**Pull request merged**